File: server/services/key_value_service.py
from storage.key_value_store import KeyValueStore
from typing import Any, List, Optional


"""
Service layer for the key-value store
"""
class KeyValueService:
    def __init__(self, store: KeyValueStore):
        self.store = store
    
    async def set_key_value(self, key: str, value: Any, ttl: Optional[int] = None) -> bool:
        return await self.store.store(key, value, ttl)
    
    async def get_value(self, key: str) -> Optional[Any]:
        return await self.store.retrieve(key)
    
    async def delete_key(self, key: str) -> bool:
        return await self.store.delete(key)
    
    async def get_all_keys(self) -> List[str]:
        return await self.store.list_keys()
    
    async def key_exists(self, key: str) -> bool:
        return await self.store.exists(key)

File: go_code/server/services/key_value_service.go
package services

import (
	"server/interfaces"
	"time"
)

// KeyValueService provides business logic for key-value operations
type KeyValueService struct {
	store interfaces.KeyValueStore
}

// NewKeyValueService creates a new key-value service
func NewKeyValueService(store interfaces.KeyValueStore) *KeyValueService {
	return &KeyValueService{
		store: store,
	}
}

// SetKeyValue sets a key-value pair with optional TTL
func (s *KeyValueService) SetKeyValue(key string, value interface{}, ttl *int) bool {
	var ttlDuration *time.Duration
	if ttl != nil {
		duration := time.Duration(*ttl) * time.Second
		ttlDuration = &duration
	}

	return s.store.Store(key, value, ttlDuration)
}

// GetValue retrieves a value by key
func (s *KeyValueService) GetValue(key string) interface{} {
	return s.store.Retrieve(key)
}

// DeleteKey deletes a key-value pair
func (s *KeyValueService) DeleteKey(key string) bool {
	return s.store.Delete(key)
}

// KeyExists checks if a key exists
func (s *KeyValueService) KeyExists(key string) bool {
	return s.store.Exists(key)
}
